clang/lib/CodeGen/CGCall.cpp

```diff
@@ -4690,8 +4690,9 @@ void CodeGenFunction::EmitCallArg(CallArgList &args, const Expr *E,
     return emitWritebackArg(*this, args, CRE);
   }
 
-  assert(type->isReferenceType() == E->isGLValue() &&
-         "reference binding to unmaterialized r-value!");
+  assert(type->isArrayParameterType() ||
+         (type->isReferenceType() == E->isGLValue()) &&
+             "reference binding to unmaterialized r-value!");
```

Comment:
I'm trying to think through how this occurs. You're somehow ending up with an expression that is an array parameter type, which is an l-value?

I assume that's only valid in the case of the out argument expression, maybe the better change is to move the assert under the next if block where we handle out argument expressions?

Reply:
I think failing the original assert means its an r-value, so its both an array parameter type and an r value?

Reply:
Can you perform the equivalent check for array parameters here, whatever that looks like?  Given a function signature, it should be possible to compute whether the argument expression is supposed to be an lvalue or rvalue.

Reply:
I think that its an LValue if its an HLSLArgOutExpr, so I did move the assert below.